### DISCUSSION TITLE
Modeling of syntactic labels

### DISCUSSION THREAD

**Discussion by gcelano** [korpling/ANNIS#878] [General]
If I have to search for all tokens with a syntactic label such as "AuxC", is the query `tok ->dep[dep_fnc=‎"AuxC‎"] tok` the only possible one? It works, but it seems that there is overhead due to the nature of the query. Should one maybe consider modeling those syntactic labels also as `<feat>`?

**Comment by thomaskrause** [comment]
Finding and filtering token can take some time because several conditions must be true ('annis:tok' label, no outgoing coverage edges). For this specific query graphANNIS already triggers an optimization that generates candidates for the source node by getting all nodes from the "dep" component (indicated by the `part-of-component-search` marker).

```
oga_v0.2.0_1> plan tok ->dep[dep_fnc="AuxC"] tok
13:04:42 [INFO] Planned query in 1 ms
oga_v0.2.0_1:
+|indexjoin (parallel) (#1 ->dep [dep_fnc="AuxC"] #2) [out: 110694, sum: 9225347, instep: 9225347]
    #1 (tok) [out: 4613074, sum: 0, instep: 0] part-of-component-search
    #2 (tok) [out: 4613074, sum: 0, instep: 0] 

```

Thus a label on the source node would probably not help that much, because there would be almost as many nodes with this label as nodes in the "dep" component. I checked if the implementation of the graph storage for the dependency component is already optimal and found an interesting part of the statistic:

```
oga_v0.2.0_1> info
Status: "fully loaded"
Token search shortcut possible: true
------------
[...]
------------
Component Pointing/dep/dep: 4369093 annnotations
Stats: nodes=4612787, avg_fan_out=0.95, max_fan_out=102, max_depth=25, tree
Implementation: PrePostOrderO32L8V1
Status: "fully loaded"
------------
```
The maximal path length `max_depth` with 25 nodes in a path seems okish, but the `max_fan_out` of 102 means there is a node with 102 outgoing dependency edges. This can influence the statistics (because graphANNIS assumes the worst case) and lead to bad decisions in the query planner.

**Reply by gcelano** [reply]
In some cases, so many dependencies is a mistake, but in others, it is not (there are about 4 sentences with a long list of names): is there a way to mitigate this issue? Do these 4 sentences make ANNIS slower?

**Reply by thomaskrause** [reply]
It depends on the specific query and could only affect queries with a more than one operator. GraphANNIS needs to decide which of the operators to use first when joining the results. It uses the statistics to estimate how many intermediate results might be generated and chooses the query plan with the least intermediate results. Thus a simple query like `tok ->dep pos="NN"` won't be affected because there is no real alternative. But if there is somethink like `tok . tok ->dep tok` a statistic that is way of can lead to slower query plans.

What I usally do is to use the `plan` subcommand in the graphANNIS CLI to see the actual execution plan for a query and the estimated intermediate result sizes. If I see that they are way of for one operator, but much more realistic for another one, it might be a possible optimization opportunity.
